File: cv/sagaSimulations/logistic_regression/logistic_regression/logistic_regression/saga.py
import numpy as np
import sys
import pkg_resources
from sklearn.metrics import log_loss


class SAGA:
    """
    Methods to apply SGLD with zero variance control variate postprocessing for logistic regression

    SGLD stands for stochastic gradient Langevin dynamics and is a MCMC method for large datasets.
    Zero variance control variates are used to improve the efficiency of the sample.

    SGLD notation used as in reference 1
    Zero variance control variate notation used as in reference 2
    References:
        1. Stochastic gradient Langevin dynamics - 
                http://people.ee.duke.edu/~lcarin/398_icmlpaper.pdf
        2. Zero variance control variates for Hamiltonian Monte Carlo - 
                https://projecteuclid.org/download/pdfview_1/euclid.ba/1393251772
    """
    
    def __init__(self,lr,epsilon,minibatch_size,n_iter):
        """
        Initialize the container for SGLD

        Parameters:
        lr - LogisticRegression object
        epsilon - the stepsize to perform SGD at
        minibatch_size - size of the minibatch used at each iteration
        n_iter - the number of iterations to perform
        """
        self.epsilon = epsilon
        # Set the minibatch size
        self.minibatch_size = minibatch_size
        self.sample_minibatch(lr)
        # Hold number of iterations so far
        self.iter = 1
        self.output = np.zeros( ( n_iter, lr.d ) )
        # Hold gradients of each data point
        self.g_alpha_i = lr.dlogdens(self,xrange(lr.N)) 
        self.g_alpha = self.g_alpha_i.sum(axis=0)


    def update(self,lr):
        """
        Update one step of stochastic gradient Langevin dynamics

        Parameters:
        lr - LogisticRegression object

        Modifies:
        lr.beta - updates parameter values using SGLD
        lr.grad_sample - adds calculated gradient to storage
        """
        self.sample_minibatch(lr)
        # Calculate gradients of log density at current point and minibatch
        dlogdensgrads_beta = lr.dlogdens(self)
        # Calculate old and new log likelihood gradient estimates
        loglikgradest_beta = dlogdensgrads_beta.sum(axis=0)
        loglikgradest_alpha = self.g_alpha_i[self.minibatch,:].sum(axis=0)
        # Calculate SAGA estimate of log posterior gradient
        dlogbeta = self.dlogpostest(lr,loglikgradest_alpha,loglikgradest_beta)

        # Update g_alpha
        self.g_alpha += loglikgradest_beta - loglikgradest_alpha
        self.g_alpha_i[self.minibatch,:] = dlogdensgrads_beta

        # Update parameters using SGLD
        eta = np.random.normal( size = lr.d, scale = self.epsilon )
        lr.beta += self.epsilon / 2 * dlogbeta + eta


    def dlogpostest(self,lr,loglikgrad_alpha,loglikgrad_beta):
        """
        Calculate SAGA minibatch estimate of gradient of the log posterior wrt the parameters

        Parameters:
        sgld - a StochasticGradientLangevinDynamics object, used to specify the minibatch

        Returns:
        dlogbeta - estimated log posterior gradient
        """
        correction = lr.N / float( self.minibatch_size )
        dlogpostest_saga = self.g_alpha + correction *( loglikgrad_beta - loglikgrad_alpha )
        # Add gradient of log prior (assume Laplace prior with scale 1)
        dlogpostest_saga -= np.sign(lr.beta)
        return dlogpostest_saga


    def sample_minibatch(self,lr):
        """Sample the next minibatch"""
        self.minibatch = np.random.choice( np.arange( lr.N ), self.minibatch_size, replace = False )

File: cv/sagaSimulations/lda/src/saga.jl
using Distributions
include("lda.jl")

"""
Container for sgld parameters
"""
type saga
    subsize::Int64                  # Minibatch size
    epsilon::Float64             # Stepsize tuning constant
    iter::Int64
    subsample::Array{Int64,1}
    g_alpha_i::Array{Float64,3}
    g_alpha::Array{Float64,2}
end

"""
Set standard tuning values
"""
function saga( model::lda, epsilon, subsize = 200::Int64 )
    iter = 1
    subsample = [-1]
    g_alpha_i = dlogdens( model, 1:model.M )
    g_alpha = squeeze( sum( g_alpha_i, 1 ), 1 )
    saga( subsize, epsilon, iter, subsample, g_alpha_i, g_alpha )
end

"""
Update one step of Stochastic Gradient Langevin Dynamics for a Latent Dirichlet Allocation model
"""
function saga_update( model::lda, tuning::saga )
    # Subsample documents
    tuning.subsample = sample( 1:model.M, tuning.subsize )
    # Calculate log density gradients for minibatch wrt theta
    dlogdens_theta = dlogdens( model, tuning.subsample )
    # Calculate minibatch log likelihood gradient estimates at theta and at alpha
    dloglik_theta = squeeze( sum( dlogdens_theta, 1 ), 1 )
    dloglik_alpha = squeeze( sum( tuning.g_alpha_i[tuning.subsample,:,:], 1 ), 1 )
    # Calculate SAGA estimate of log posterior gradient
    dlogpostest = dlogpostsaga( model, tuning, dloglik_theta, dloglik_alpha )
    if tuning.iter % 10 == 0
        println(sum(dlogpostest))
    end

    # Update g_alpha
    tuning.g_alpha += dloglik_theta - dloglik_alpha
    tuning.g_alpha_i[tuning.subsample,:,:] = dlogdens_theta

    # Gibbs update of z topic allocations
    model.zcounts[tuning.subsample,:,:] = update_topics( model, tuning.subsample )
    # Simulate Langevin dynamics using SAGA log posterior estimate
    injected_noise = Normal( 0, 1 )
    model.theta += tuning.epsilon/2 * dlogpostest
    model.theta += sqrt(tuning.epsilon) * rand( injected_noise, ( model.K, model.V )  )
end

"""
Calculate SAGA log posterior estimate at current state model.theta
"""
function dlogpostsaga( model::lda, tuning::saga, dloglik_theta::Array{Float64,2}, 
        dloglik_alpha::Array{Float64,2} )
    correction = model.M / tuning.subsize
    dlogpostest = tuning.g_alpha + correction * ( dloglik_theta - dloglik_alpha )
    dlogpostest += dlogprior( model )
    return dlogpostest
end
